### ISSUE TITLE
Drag-and-drop from an X11 FLTK app to a Wayland app does not work.

### ISSUE THREAD

**Issue by ManoloFLTK** [fltk/fltk#519]
Drag-and-drop from an X11 FLTK app to a Wayland app (FLTK-based or not) does not work.

**To Reproduce**
Run under Linux 2 instances of test/editor, one using Wayland, one using X11.
Put some text in the X11 instance. Select some text. Drag it to the Wayland editor window:
no text is transferred.

1. Buid FLTK with -DOPTION_USE_WAYLAND=1
2. Type,  in the build directory
./bin/test/editor &
FLTK_BACKEND=x11  ./bin/test/editor &
4. Attempt to drag text from the X11 window to the Wayland window

**Expected behavior**
Text should be copied from one window to the other. That does not occur.

**FLTK Version**
 - Version:  1.4.0
 - Downloaded and self-built from:  Git commit: 24c63c0

**FLTK Configure / Build Options**
  ./configure --enable-wayland

**Operating System / Platform:**
 - OS: Linux
 - OS Version: Ubuntu 22.04

**Additional context**
1) The converse operation, dragging from Wayland to X11, works correctly.
2) Dragging from a non-FLTK X11 app (e.g. gedit launched by `WAYLAND_DISPLAY=NONE gedit &` )
to a Wayland app (FLTK-based or not) does work.
3) Dragging from X11/FLTK to a non-FLTK Wayland app, e.g. gedit, doesn't work.

This suggests that something is missing in the X11 leg of libfltk that informs Wayland apps that a new "Wayland data_offer" is being proposed by the X11 app.

**Comment by Albrecht-S** [comment]
@ManoloFLTK Here's an idea how to test this, however I tried it and I don't see what's going on (i.e. right or wrong).
1. Launch a Wayland app, for instance `gedit`
2. launch a working X11 app, for instance (I used) `WAYLAND_DISPLAY=NONE xed`
3. run in a terminal: `xwininfo` and click on the X11 window to get the window id
4. run in a terminal: `xev -id <window id from (3)>`
5. view and record events when dragging from X11 to Wayland
6. launch a FLTK/X11 app and repeat points 3-5 to find out what's different.

A quick test didn't reveal anything useful for me though, but maybe you can find something, particularly if you also use a FLTK/Wayland app as target and "watch" Wayland events (in a debugger or by using printf's etc.).

That's just an idea...

**Comment by ManoloFLTK** [comment]
@Albrecht-S : What is xed ? To what package does it belong ?

**Comment by Albrecht-S** [comment]
`xed` is a text editor installed on Linux Mint. Its package is `xed`. `apt install xed`. It's just another pretty simple text editor.

**Comment by ManoloFLTK** [comment]
@Albrecht-S : Here is the result of some exploration.

1) Experiment in Ubuntu 22.04 under the gnome desktop:
```
kwrite &                 <--- starts as X11 client
gedit&                    <--- starts as Wayland client
Drag from kwrite   <--- cursor shows dragged text
Drop to gedit         <--- does not work
```

kwrite is part of the KDE desktop and is not linked to libgdk-*.so
Interestingly, kwrite starts as an X11 client when run on the gnome desktop
although it's linked to libwayland-client.so and it starts as a Wayland client 
on the KDE desktop.

2) Experiment with
```
gedit&                               <--- starts as Wayland client
WAYLAND_DISPLAY=NONE gdb nautilus  <--- starts as X11 client
break gtk_drag_begin_with_coordinates
break gdk_drag_context_manage_dnd
break gdk_drag_context_get_protocol
run
```
and then drag a small file from nautilus to gedit.
We see these GTK and GDK functions get called by the nautilus X11 client :
- when the drag operation begins:
`gtk_drag_begin_with_coordinates() `in libgtk-3.so
then
`gdk_drag_context_manage_dnd()` in libgdk-3.so
- later, when dropping to gedit Wayland window
`gdk_drag_context_get_protocol()` in libgdk-3.so

Altogether, this suggests that DnD from X11 to Wayland is managed inside libgdk-3.so
and that it requires a GTK window as source and a GTK window as target.
The X11 FLTK platform cannot provide GTK windows.

In conclusion, making FLTK support DnD from X11 to Wayland probably requires a lot of new code
that would duplicate what libgdk.so does.

**Comment by Albrecht-S** [comment]
@ManoloFLTK I see your test results but I'm not sure your conclusion is correct [1]. But anyway, we should probably not invest too much time if other X applications can't interact with Wayland as well.

I've also noticed that `VS Code` (an X application) doesn't work with other Wayland **and** FLTK/X11 apps (WRT DND). This may, however, be caused by `VS Code` running as a `snap` rather than a native app. I don't know...

[1] In my - maybe too naïve - imagination all (not only FLTK) X applications should interact somehow with XWayland, the Wayland X server, and XWayland with the Wayland application, or something like that. I wonder why it is so problematic though. Maybe a flaw in XWayland? Or maybe I'm totally wrong with my assumption.

**Comment by ManoloFLTK** [comment]
@Albrecht-S  XWayland is a client of the Wayland compositor, just like any other Wayland app. So, XWayland can communicate with the compositor but not with any other Wayland client, since clients are completely isolated from each other in the Wayland model. DnD from X11 to Wayland would require to ask XWayland to behave as a Wayland client app initiating a DnD and to feed this operation with the X11 dragged data. That is not part of the X protocol. I don't know if there's an API or protocol somewhere to do that.

**Comment by Albrecht-S** [comment]
Manolo, thanks for your explanation, it looks more complicated than I had thought.

**Comment by ManoloFLTK** [comment]
@Albrecht-S  I've found [there](https://gitlab.gnome.org/GNOME/mutter/-/issues/2136) the attached pure-X demo program that does perform DnD from itself (X11) to FLTK-Wayland. Thus, I just have to find out what it does that FLTK-X11 doesn't do.

[XDnDSource.c.txt](https://github.com/fltk/fltk/files/9846043/XDnDSource.c.txt)

**Comment by ManoloFLTK** [comment]
@Albrecht-S It turns out the reason why FLTK-X11 apps don't perform DnD to Wayland apps is in function `Fl_X11_Screen_Driver::dnd(int unused)` of file `fl_dnd_x.cxx`. Statement at line 88
  `XSetSelectionOwner(fl_display, fl_XdndSelection, fl_message_window, fl_event_time);`
should be changed to
  `XSetSelectionOwner(fl_display, fl_XdndSelection, source_window, fl_event_time);`
That is, the owner of the X selection should be set to the true window `source_window`rather than the dummy, invisible `fl_message_window` whose utility I don't grasp now.

Since you've been working on X selection stuff recently when dealing with large data transfers, do you have an opinion about whether there could be unwanted side effects of the proposed change? I've tested the modified code here and find that DnD operations of UTF8 text within X11, from X11 to Wayland, and from Wayland to X11 all work OK.

**Comment by ManoloFLTK** [comment]
@Albrecht-S I see your  last message after having written mine.
The problem you mention no longer exists with Ubuntu 22.04 which gives
```
gnome-shell --version
GNOME Shell 42.4
```

**Comment by Albrecht-S** [comment]
> The problem you mention no longer exists with Ubuntu 22.04

I tested with my Debian Bullseye VM (`GNOME Shell 3.38.6`) and this still shows the issue. This system has not been updated for a while though. But maybe it's fixed in the next Debian release.

@ManoloFLTK You wrote above:
> Statement at line 88
> `  XSetSelectionOwner(fl_display, fl_XdndSelection, fl_message_window, fl_event_time);`
> should be changed to
> `XSetSelectionOwner(fl_display, fl_XdndSelection, source_window, fl_event_time);`
> ... do you have an opinion about whether there could be unwanted side effects of the proposed change?

Off the top of my head I don't know what `fl_message_window` is good for or what value it got assigned when DnD is active. I'll look into it and report here what I find out.

Anyway, it's great that you found a way to change the FLTK code to make it work. That means - if I understand correctly - if your change is good (i.e. w/o unwanted side effects) we have a way to change it right now w/o depending on other Wayland related changes. This sounds very good.

**Comment by ManoloFLTK** [comment]
@Albrecht-S I've looked at the use in FLTK of the dummy, invisible, X11 window corresponding to global variable `fl_message_window`. I think the case for not using it while starting a DnD operation is very clear because starting a DnD is possible only when a true X11 window is mapped, so there's no reason to use a fake window for the `XSetSelectionOwner()` call mentionned above. I will therefore commit the change to this `XSetSelectionOwner()` call and close this issue. If you were to find a possible unwanted side effect of this change, we could take care of it as a second issue.

**Comment by Albrecht-S** [comment]
OK, thanks.
